### ISSUE TITLE
Ability to call an external API with template variables

### ISSUE THREAD

**Issue by regicsolutions** [cookiecutter/cookiecutter#1494]
* Cookiecutter version: latest
* Template project url: N/A
* Python version: 3.8
* Operating System: Win 10

### Description:

I cant find any documentation regarding calling an external API using the CLI/post_gen hook in either py or sh script. I was looking to call/pass template variables to a lambda function. Is this something cookiecutter supports out of the box?

### What I've run:

```
// REPLACE ME: Paste a log of command(s) you ran and cookiecutter's output, tracebacks, etc, here
```

**Comment by edesz** [comment]
Hi @regicsolutions, per the [docs](https://cookiecutter.readthedocs.io/en/latest/advanced/hooks.html#template-variables-are-rendered-in-the-script), this is supported in `cookiecutter`.

Here is a basic example to call the [WorldTime API](http://worldtimeapi.org/) using a post-gen hook (Python) using a template variable.

Directory structure for `cookiecutter-demo/`
```bash
├── cookiecutter.json
├── {{cookiecutter.project_name}}
│   └── mytest.txt
└── hooks
    └── post_gen_project.py
```

`cookiecutter.json` has two variables
```
{
    "timezone": "Europe/Madrid",
    "project_name": "myproject"
}
```

The post-gen hook is `hooks/post_gen_project.py` and it uses `jinja` syntax to access the template variable `{{ cookiecutter.timezone }}`
```python
import requests

timezone = "{{ cookiecutter.timezone }}"

url = f"http://worldtimeapi.org/api/timezone/{timezone}"
r = requests.get(url)
current_time = r.json()["datetime"]
print(f"The current time in the {timezone} timezone is {current_time}")
```

`{{cookiecutter.project_name}}/mytest.txt` contains the following
```
The specified timezone is {{cookiecutter.timezone}}.
```

When this is run using
```bash
$ cookiecutter cookiecutter-demo
```
using default values at the prompts
```bash
timezone [Europe/Madrid]: 
project_name [myproject]: 
```

it produces the `myproject/mytest.txt` with the following contents
```
The specified timezone is Europe/Madrid. 
```

and the hook prints the following to the screen
```
The current time in the Europe/Madrid timezone is 2021-01-29T21:12:52.758029+01:00
```

Hope this helps to resolve your issue.

**Comment by regicsolutions** [comment]
@edesz Thanks a lot! Exactly what I was looking for

**Comment by edesz** [comment]
@regicsolutions that's great! Glad it helped out.:satisfied: This is a really powerful feature of `cookiecutter`. Template variables can also be accessed in pre-gen hooks.

Feel free to close this issue if its been resolved. :lock:

**Comment by regicsolutions** [comment]
Besides importing 'requests' which I couldn't find in the documentation  are there other options available within the cookiecutter CLI? A few other things I was looking to do as a post-hook were: 

1. Create a branch and open a pull request in Bitbucket.
2. Invoke a lambda function which requires AWS authentication.

**Comment by edesz** [comment]
Regarding `requests`, that is specific to the hook script I was demonstrating - I only needed it since I wanted to query the World Time API. So, it wouldn't be in the `cookiecutter` docs. The functionality I implemented in the hook is independent of `cookiecutter` (which only generates the templates). So, I had to import `requests` only to implement the functionality I wanted in the hook script. `cookiecutter` just provides the ability to use such hooks but implementing such custom functionality (inside the hook script) is up to the user.

Those tasks you are referring to are unfortunately beyond anything that `cookiecutter` is designed to do.

The post gen hook I used above was in Python, but it could also be a shell script `post_gen_project.sh`. If you want to do those other things in a post-gen hook script then, in that script (Python or shell), you would need to write code to
1. create a bitbucket branch ([maybe this gets you started?](https://support.atlassian.com/bitbucket-cloud/docs/branch-a-repository/#Creating-a-branch)) and issue a pull request
2. authenticate for an AWS Lambda (something like [this](https://aws.amazon.com/blogs/compute/easy-authorization-of-aws-lambda-functions/)?)

Those tasks you're referring to are specific to the use you case you're after. Being a generic template creator, `cookiecutter` doesn't offer support for coding hooks - it only ensures that the template variables are available in those hooks, but the user is responsible for building up the hook script as required. That means you'll have to find a way to implement those options (branch, pull request, Lambda authentication) using CLI commands (shell) or in Python.
